Python/ceval.c

```diff
@@ -2903,6 +2903,19 @@ _PyEval_ImportFrom(PyThreadState *tstate, PyObject *v, PyObject *name)
         }
         else {
             assert(rc == 0);
+            if (origin == NULL) {
```

Comment:
Why not do this immediately after `_PyModuleSpec_GetFileOrigin()`?

Reply:
There are a few `assert(origin);` in this function.  If we do leave this here, it would probably make sense to fix those in this PR.

Reply:
Great question!

Basically, I thought the cases where `module.__spec__.has_location = False` would be sort of weird. We expect the value we pass to `_PyModule_IsPossiblyShadowing` to be a file system path and I think `__file__` can sometimes be weird, e.g. zipped modules.

https://github.com/python/cpython/blob/aea0c586d181abb897511b6b46d28bfbe4858f79/Objects/moduleobject.c#L845-L847

The `assert(origin)` in this function are correct asserts. `_PyModule_IsPossiblyShadowing` returns 0 if origin is NULL

Reply:
Then maybe call it after `_PyModule_IsPossiblyShadowing`? There is a code above for raising exception for non-shadowed initialized module which uses `origin`.